compendium/ConfigurationAdmin/src/CMActivator.cpp

```diff
@@ -28,155 +28,167 @@
-                "Failed to create CMBundleExtension for " +
-                  bundle.GetSymbolicName(),
-                std::current_exception());
-  }
-}
-
-void CMActivator::RemoveExtension(const cppmicroservices::Bundle& bundle)
-{
-  auto const& headers = bundle.GetHeaders();
-  // bundle has no "cm" configuration
-  if (headers.find(CMConstants::CM_KEY) == std::end(headers)) {
-    logger->Log(SeverityLevel::LOG_DEBUG,
-                "No CM Configuration found in bundle " +
-                  bundle.GetSymbolicName());
-    return;
-  }
-
-  bool extensionFound = false;
-  {
-    std::lock_guard<std::mutex> l(bundleRegMutex);
-    auto extensionIt = bundleRegistry.find(bundle.GetBundleId());
-    if (extensionIt != std::end(bundleRegistry)) {
-      bundleRegistry.erase(extensionIt);
-      extensionFound = true;
-    }
-  }
-  if (extensionFound) {
-    logger->Log(SeverityLevel::LOG_DEBUG,
-                "Removed CMBundleExtension for " + bundle.GetSymbolicName());
-    return;
-  }
-  logger->Log(SeverityLevel::LOG_DEBUG,
-              "Found no CMBundleExtension for " + bundle.GetSymbolicName());
-}
-
-void CMActivator::BundleChanged(const cppmicroservices::BundleEvent& evt)
-{
-  auto bundle = evt.GetBundle();
-  const auto eventType = evt.GetType();
-  if (bundle ==
-      runtimeContext.GetBundle()) // skip events for this (runtime) bundle
-  {
-    return;
-  }
-
-  // TODO: revisit to include LAZY_ACTIVATION when supported by the framework
-  if (cppmicroservices::BundleEvent::BUNDLE_STARTED == eventType) {
-    CreateExtension(bundle);
-  } else if (cppmicroservices::BundleEvent::BUNDLE_STOPPING == eventType) {
-    RemoveExtension(bundle);
-  }
-  // else ignore
-}
-} // cmimpl
-} // cppmicroservices
-
-CPPMICROSERVICES_EXPORT_BUNDLE_ACTIVATOR(
-  cppmicroservices::cmimpl::CMActivator) // NOLINT
+        void
+        CMActivator::Start(BundleContext context)
+        {
+            runtimeContext = context;
+            // Create the Logger object used by this runtime
+            logger = std::make_shared<CMLogger>(context);
+            logger->Log(SeverityLevel::LOG_DEBUG, "Starting CM bundle");
+            // Create the AsyncWorkService object used by this runtime
+            asyncWorkService = std::make_shared<CMAsyncWorkService>(context, logger);
+            // Create ConfigurationAdminImpl
+            configAdminImpl = std::make_shared<ConfigurationAdminImpl>(runtimeContext, logger, asyncWorkService);
+            // Add bundle listener
+            bundleListenerToken
+                = context.AddBundleListener(std::bind(&CMActivator::BundleChanged, this, std::placeholders::_1));
+            // HACK: Workaround for lack of Bundle Tracker. Iterate over all bundles and call the tracker method
+            // manually
+            for (auto const& bundle : context.GetBundles())
+            {
+                if (cppmicroservices::Bundle::State::STATE_ACTIVE == bundle.GetState())
```

Comment:
I don't know if this is a real problem or not, but the State enumeration is definitely a uint32_t with different bits representing different states. If it's ever possible to have more than one state bit set at a time, this equality check can fail. If it's NOT possible to have more than one state bit set at a time, why is the enum a bunch of bit flags?

This could really be a problem. Here's the comment from Bundle.h regarding this very thing:

* Values assigned to these states have no specified ordering; they represent
 * bit values that may be ORed together to determine if a bundle is in one of
 * the valid states.

Reply:
@carneyweb please create a github issue for this